modules/host/update.go

```diff
@@ -313,7 +295,25 @@ func (h *Host) ProcessConsensusChange(cc modules.ConsensusChange) {
 		h.log.Println(err)
 	}
 
-	// Handle the list of action items.
+	// Handle the list of action items. Action items require host locks and
+	// potentially require waiting for long periods of time while various
+	// network communications finish. To prevent the host lock from being held
+	// while the network I/O is blocking the action item execution, the action
+	// items are handled in their own thread. A wait group is used to prevent
+	// the host from releasing ThreadGroup control until all threads that this
+	// function have spawned have completed.
+	wg := new(sync.WaitGroup)
+	// The host should not release thread group control until all spawned
+	// threads have completed execution, though waiting for them should not
+	// block this function. The host should also not release thread group
+	// control until this function has terminated, hence the use of 'defer
+	// func() { go func() }' to coordinate waiting and done-ing.
+	defer func() {
+		go func() {
+			wg.Wait()
```

Comment:
Yeah it's just weird because you don't want to block the ProcessConsensusChange function while it's child threads are running (which is part of the reason they are child threads). In terms of closeness, you are still doing `Add; defer Done`, it's just `Add; defer go Done`. It allows you to cleanly encapsulate a bunch of child threads into the same thread group without needing the parent thread to outlive all of the children.

imo it's good design to have concurrency where your parent doesn't need to stick around. It's a little different than how most people approach the problem but I think it allows you to approach a broader range of problems in a more natural way.